The library is not compatible with TypeScript 4.4

**What is the current behavior?**

**Steps to reproduce it and if possible a minimal demo of the problem. Your bug will get fixed much faster if we can run your code and it doesn't have extra dependencies other than `react-use`. Paste the link to your [JSFiddle](https://jsfiddle.net) or [CodeSandbox](https://codesandbox.io) example below:**

1. Install latest TypeScript (`npm install typescript@beta`)
2. Build your project (make sure `skipLibCheck` is non-existent or `false` in your `tsconfig.json`)

The build will fail with the following errors in `node_modules/react-use/lib/usePermission.d.ts`.

- Cannot find name 'DevicePermissionDescriptor'.
- Cannot find name 'MidiPermissionDescriptor'.
- Cannot find name 'PushPermissionDescriptor'.

**What is the expected behavior?**

Project builds.

**A little about versions:**  
- _OS_: n/a
- _Browser (vendor and version)_:  n/a
- _React_: n/a
- _`react-use`_: `17.2.4`
- _Did this worked in the previous package version?_ In TypeScript 4.3, yes.

TypeScript 4.4 introduces breaking changes to `lib.d.ts`. Here's [the full list](https://devblogs.microsoft.com/typescript/announcing-typescript-4-4-beta/). These interfaces are no longer part of the standard library.

I searched for the missing types in the [TypeScript repo](https://github.com/microsoft/TypeScript) and found them in `lib/lib.webworker.d.ts`. So my immediate thought was to make this tweak in the project’s `tsconfig.json`:

```diff
-   "lib": ["es2020", "dom"],
+   "lib": ["es2020", "dom", "webworker"],
```

Unfortunately, this change introduced dozens of new errors like:

```
node_modules/typescript/lib/lib.dom.d.ts:25:1 - error TS6200: Definitions of the following identifiers conflict with those in another file: EventListenerOrEventListenerObject, ImportExportKind, TableKind, ValueType, ExportValue, Exports, ImportValue, ModuleImports, Imports, name, HeadersInit, BodyInit, RequestInfo, BlobPart, DOMHighResTimeStamp, CanvasImageSource, MessageEventSource, ImageBitmapSource, OnErrorEventHandler, TimerHandler, PerformanceEntryList, ReadableStreamReader, ReadableStreamController, VibratePattern, AlgorithmIdentifier, HashAlgorithmIdentifier, BigInteger, NamedCurve, GLenum, GLboolean, GLbitfield, GLint, GLsizei, GLintptr, GLsizeiptr, GLuint, GLfloat, GLclampf, TexImageSource, Float32List, Int32List, GLint64, GLuint64, Uint32List, BufferSource, DOMTimeStamp, FormDataEntryValue, IDBValidKey, Transferable, ReadableStreamDefaultReadResult, BinaryType, CanvasDirection, CanvasFillRule, CanvasLineCap, CanvasLineJoin, CanvasTextAlign, CanvasTextBaseline, ClientTypes, ColorSpaceConversion, EndingType, IDBCursorDirection, IDBRequestReadyState, IDBTransactionMode, ImageOrientation, ImageSmoothingQuality, KeyFormat, KeyType, KeyUsage, NotificationDirection, NotificationPermission, PermissionName, PermissionState, PremultiplyAlpha, PushEncryptionKeyName, PushPermissionState, ReferrerPolicy, RequestCache, RequestCredentials, RequestDestination, RequestMode, RequestRedirect, ResizeQuality, ResponseType, ServiceWorkerState, ServiceWorkerUpdateViaCache, VisibilityState, WebGLPowerPreference, WorkerType, XMLHttpRequestResponseType
node_modules/typescript/lib/lib.dom.d.ts:3984:5 - error TS2374: Duplicate index signature for type 'number'.
node_modules/typescript/lib/lib.dom.d.ts:5227:5 - error TS2374: Duplicate index signature for type 'number'.
...
```

So clearly that’s not the right way forward 😅

--- a/src/usePermission.ts
+++ b/src/usePermission.ts
@@ -1,45 +1,59 @@
 import { useEffect, useState } from 'react';
 import { noop, off, on } from './misc/util';
 
-type PermissionDesc =
+export type IState = PermissionState | '';
+
+interface IPushPermissionDescriptor extends PermissionDescriptor {
+  name: 'push';
+  userVisibleOnly?: boolean;
+}
+
+interface IMidiPermissionDescriptor extends PermissionDescriptor {
+  name: 'midi';
+  sysex?: boolean;
+}
+
+interface IDevicePermissionDescriptor extends PermissionDescriptor {
+  name: 'camera' | 'microphone' | 'speaker';
+  deviceId?: string;
+}
+
+export type IPermissionDescriptor =
   | PermissionDescriptor
-  | DevicePermissionDescriptor
-  | MidiPermissionDescriptor
-  | PushPermissionDescriptor;
+  | IPushPermissionDescriptor
+  | IMidiPermissionDescriptor
+  | IDevicePermissionDescriptor;
 
-type State = PermissionState | '';
-
-const usePermission = (permissionDesc: PermissionDesc): State => {
-  let mounted = true;
-  let permissionStatus: PermissionStatus | null = null;
-
-  const [state, setState] = useState<State>('');
-
-  const onChange = () => {
-    if (mounted && permissionStatus) {
-      setState(permissionStatus.state);
-    }
-  };
-
-  const changeState = () => {
-    onChange();
-    on(permissionStatus, 'change', onChange);
-  };
+// const usePermission = <T extends PermissionDescriptor>(permissionDesc: T): IState => {
+const usePermission = (permissionDesc: IPermissionDescriptor): IState => {
+  const [state, setState] = useState<IState>('');
 
   useEffect(() => {
+    let mounted = true;
+    let permissionStatus: PermissionStatus | null = null;
+
+    const onChange = () => {
+      if (!mounted) {
+        return;
+      }
+      setState(() => permissionStatus?.state ?? '');
+    };
+
     navigator.permissions
       .query(permissionDesc)
       .then((status) => {
         permissionStatus = status;
-        changeState();
+        on(permissionStatus, 'change', onChange);
+        onChange();
       })
       .catch(noop);
 
     return () => {
-      mounted = false;
       permissionStatus && off(permissionStatus, 'change', onChange);
+      mounted = false;
+      permissionStatus = null;
     };
-  }, []);
+  }, [permissionDesc]);
 
   return state;
 };

--- a/src/useSpeech.ts
+++ b/src/useSpeech.ts
@@ -1,54 +1,90 @@
-import { useRef } from 'react';
-import useMount from './useMount';
-import useSetState from './useSetState';
-import { isBrowser } from './misc/util';
+import { useCallback, useEffect, useRef, useState } from 'react';
 
-export interface SpeechState {
-  isPlaying: boolean;
+type SpeechOptions = {
   lang: string;
-  voice: SpeechSynthesisVoice;
+  voice?: SpeechSynthesisVoice;
   rate: number;
   pitch: number;
   volume: number;
+};
+
+export type ISpeechOptions = Partial<SpeechOptions>;
+
+export type VoiceInfo = Pick<SpeechSynthesisVoice, 'lang' | 'name'>;
+
+export type ISpeechState = SpeechOptions & {
+  isPlaying: boolean;
+  status: string;
+  voiceInfo: VoiceInfo;
+};
+
+enum Status {
+  init,
+  play,
+  pause,
+  end,
 }
 
-export interface SpeechOptions {
-  lang?: string;
-  voice?: SpeechSynthesisVoice;
-  rate?: number;
-  pitch?: number;
-  volume?: number;
-}
-
-const voices =
-  isBrowser && typeof window.speechSynthesis === 'object' ? window.speechSynthesis.getVoices() : [];
-
-const useSpeech = (text: string, opts: SpeechOptions = {}): SpeechState => {
-  const [state, setState] = useSetState<SpeechState>({
-    isPlaying: false,
-    lang: opts.lang || 'default',
-    voice: opts.voice || voices[0],
-    rate: opts.rate || 1,
-    pitch: opts.pitch || 1,
-    volume: opts.volume || 1,
+const useSpeech = (text: string, options: ISpeechOptions): ISpeechState => {
+  let mounted = useRef<boolean>(false);
+  const [state, setState] = useState<ISpeechState>(() => {
+    const { lang = 'default', name = '' } = options.voice || {};
+    return {
+      isPlaying: false,
+      status: Status[Status.init],
+      lang: options.lang || 'default',
+      voiceInfo: { lang, name },
+      rate: options.rate || 1,
+      pitch: options.pitch || 1,
+      volume: options.volume || 1,
+    };
   });
 
-  const uterranceRef = useRef<SpeechSynthesisUtterance | null>(null);
+  const handlePlay = useCallback(() => {
+    if (!mounted.current) {
+      return;
+    }
+    setState((preState) => {
+      return { ...preState, isPlaying: true, status: Status[Status.play] };
+    });
+  }, []);
 
-  useMount(() => {
+  const handlePause = useCallback(() => {
+    if (!mounted.current) {
+      return;
+    }
+    setState((preState) => {
+      return { ...preState, isPlaying: false, status: Status[Status.pause] };
+    });
+  }, []);
+
+  const handleEnd = useCallback(() => {
+    if (!mounted.current) {
+      return;
+    }
+    setState((preState) => {
+      return { ...preState, isPlaying: false, status: Status[Status.end] };
+    });
+  }, []);
+
+  useEffect(() => {
+    mounted.current = true;
     const utterance = new SpeechSynthesisUtterance(text);
-    opts.lang && (utterance.lang = opts.lang);
-    opts.voice && (utterance.voice = opts.voice);
-    utterance.rate = opts.rate || 1;
-    utterance.pitch = opts.pitch || 1;
-    utterance.volume = opts.volume || 1;
-    utterance.onstart = () => setState({ isPlaying: true });
-    utterance.onresume = () => setState({ isPlaying: true });
-    utterance.onend = () => setState({ isPlaying: false });
-    utterance.onpause = () => setState({ isPlaying: false });
-    uterranceRef.current = utterance;
-    window.speechSynthesis.speak(uterranceRef.current);
-  });
+    options.lang && (utterance.lang = options.lang);
+    options.voice && (utterance.voice = options.voice);
+    utterance.rate = options.rate || 1;
+    utterance.pitch = options.pitch || 1;
+    utterance.volume = options.volume || 1;
+    utterance.onstart = handlePlay;
+    utterance.onpause = handlePause;
+    utterance.onresume = handlePlay;
+    utterance.onend = handleEnd;
+    window.speechSynthesis.speak(utterance);
+
+    return () => {
+      mounted.current = false;
+    };
+  }, []);
 
   return state;
 };

--- a/stories/useSpeech.story.tsx
+++ b/stories/useSpeech.story.tsx
@@ -3,9 +3,8 @@ import * as React from 'react';
 import { useSpeech } from '../src';
 import ShowDocs from './util/ShowDocs';
 
-const voices = window.speechSynthesis.getVoices();
-
 const Demo = () => {
+  const voices = window.speechSynthesis.getVoices();
   const state = useSpeech('Hello world!', { rate: 0.8, pitch: 0.5, voice: voices[0] });
 
   return <pre>{JSON.stringify(state, null, 2)}</pre>;

